File: Scaler/SortedPermutationRank.cpp
#include <bits/stdc++.h>
// #include <string.h>

class Solution {
    Solution();
    int findRank(std::string A);
};

long fact(int x) {
    return x < 3 ? x : x*fact(x-1);
}

void update(int count[], char c) {
    for (int i=c; i<256; i++) count[i]--;
}
#define mod 1000003

int Solution::findRank(std::string A) {
    int count[256] = {0};
    int len = A.length();
    // char arr[len+1];
    // for (int)
    for (int i=0; i<len; i++) count[A[i]]++;
    for (int i=1; i<256; i++) count[i] += count[i-1];
    
    int rank=1;
    long perms=fact(len);
    
    for(int i=0; i<len; i++) {
        perms /= len - i;
        rank += (perms*count[A[i] - 1])%mod;
        update(count, A[i]);
    }
    return rank%mod;
}

File: Scaler/SortedPermutationRank.java
import java.util.Arrays;

public class SortedPermutationRank {

    // O(n) runtime, O(256) or O(1) space
    static int findRank(String str) {
        int[] count= new int[256];
        int rank = 1;
        Arrays.fill(count, 0);
        char[] arr = str.toCharArray();
        int len = arr.length;
        long perms = factorial(len);
        // System.out.println(perms);

        // Count Array
        for (int i=0; i<len; i++) count[arr[i]]++;
        for (int i=1; i<256; i++) count[i] += count[i-1];
        // System.out.println(Arrays.toString(count));

        for (int i=0; i<len; i++) {
            perms /= len - i;
            rank += count[arr[i] - 1] * perms;
            updateCount(count, arr[i]);
        }
        // return rank; 
        return rank%1000003; //or for huge int > 1000003
    }

    static long factorial(long n) {
        if (n < 0) return -1;
        return (n < 2) ? 1 : n*factorial(n-1); 
    }

    static void updateCount(int[] count, char ch) {
        for (int i=ch; i<256; i++) --count[i];
    }

    // O(n^2)
    static long findLexRank(String str) {
        char[] arr = str.toCharArray();
        int len = arr.length; 
        long rank = 1;
        for (int i=0; i<len; i++) {
            int count=0;
            for (int j=i+1; j<len; j++) {
                if (arr[i] > arr[j]) count++;
            }
            rank += count*factorial(len - 1 - i);
        }
        return rank%1000003;
    }

    public static void main(String[] args) {
        String str = "DTNGJPURFHYEW";
        String str2 = "ZCSFLVHXRYJQKWABGT";
        System.out.println(findRank(str));
        System.out.println(findLexRank(str2));
    }
}
